java-client/src/main/java/org/opensearch/client/util/PathEncoder.java

```diff
@@ -0,0 +1,89 @@
+package org.opensearch.client.util;
```

Comment:
This needs a license header.

If the code is copied from another APLv2 project, include their license header and a clear reference, too.

Reply:
@dblock I don't feel right to copy the code like that: at the moment, we always have dependency on either AHC 4 or AHC 5, we could make `PathEncoder` to detect which one is on classpath and use the respective `URLEncodedUtils` .

Reply:
We certainly have the right to copy code per license, but maybe the right solution is to refactor in a way that the implementation is not in `java-client/src/main/java/org/opensearch/client/transport/endpoints/SimpleEndpoint.java`, but in a version-specific transport implementation? @AntCode97 WDYT?

Reply:
> We certainly have the right to copy code per license,

Sure, no issues with that but maintaining it is not something we should take lightly

Reply:
If we're going to continue to have dependencies on AHC, I think it's better to use URLEncodedUtils per version, with PathEncoder detecting the classpath, because implementing it would be difficult to maintain, and if a security vulnerability is found, it can't be fixed by simply raising the library version.

Reply:
> If we're going to continue to have dependencies on AHC, I think it's better to use URLEncodedUtils per version, 

@AntCode97 could you update your pull request to use take this path? thank you

Reply:
Okay, but I'm busy today, so I'll update it next week.

Reply:
Please take our own pace, no rush, thank you!

Reply:
While working on this, it occurred to me that modifying the SimpleEndpoint class in the 2.x branch to import org.apache.hc.core5.net.URLEncodedUtils would solve the problem.
This is because the main branch uses AHC5 and the 2.x branch uses both AHC4 and AHC5.

The opensearch-rest-client uses AHC4, but I don't think it would be a problem to change everything else to AHC5.

What do you think? @reta  If you agree, I'm going to close this PR, create a new branch based on the 2.x branch, and open the PR with the modification to import org.apache.hc.core5.net.URLEncodedUtils.

Reply:
> While working on this, it occurred to me that modifying the SimpleEndpoint class in the 2.x branch to import org.apache.hc.core5.net.URLEncodedUtils would solve the problem.

This is the large difference between 2.x and 3.x (main): 2.x uses by default RestClient and AHC 4.x, whereas 3.x by default uses AHC 5.x all the way. We cannot use AHC 5.x in 2.x, its presence is optional.

Reply:
I got it. I will open PR soon

Reply:
Okay. So, since 3.x uses AHC5 by default, I assume that AHC 4 and AHC 5 can exist optionally in 2.x only.
If so, should I close this PR and create a new branch based on the 2.x branch, detect classpasth, and open a PR that uses URLEncodedUtils per version?

Reply:
I found a simpler way. 
It is to utilize java.net.URLEncoder. 
When I encode using URLEncoder, it passes the EndpointTest, so it looks like it's okay to use it instead of URLEncodedUtils.
I've fixed it in the code.

Reply:
If you're okay with using`URLEncoder`, I'll open a PR to make the same change to the 2.x branch.
This will also cause a modification to `EndpointTest` in the 2.x branch, since `EndpointTest` has tests written based on the `org.apache.http.client.utils.URLEncodedUtils` class.

Reply:
For 2.x we have a backport process, once this is merged it will attempt to automatically backport.

Reply:
@AntCode97 @dblock this is very risky change, one of the reason why others (like AHC, ...) have to reinvent this URL encoding / decoding / parsing utils is coming from the fact that Java's URL implementation is not compliant with RFC-3986 [1]. The consequences of that is - many edge (and no so edge) cases are not covered, and weird URLs are accepted where they should not. The AHC 5.x seems to keep conformance [2].

[1] https://datatracker.ietf.org/doc/html/rfc3986
[2] https://github.com/apache/httpcomponents-core/pull/205

Reply:
@AntCode97 Thanks so much for hanging in here with us for what is seemingly a simple change. It's really not! Appreciate your time and work.

Reply:
After performing a clean build, I was unable to build.
The reason is that I can't get `org.apache.lucene:lucene-test-framework:9.11.0-snapshot-fb97840`, which is used by `org.opensearch.test:framework:3.0.0-SNAPSHOT`.
I had no problem until yesterday, but suddenly this happened. @reta, do you know anything about this?

Reply:
> I had no problem until yesterday, but suddenly this happened. @reta, do you know anything about this?

@AntCode97 yes, I do, but I am very much surprised you are impacted. This is by and large "normal" dependency update cycle (we bump Apache Lucene snapshots). The repo URLs are coming from OpenSearch Gradle plugin(s) which should fetch the artifacts transparently but it seems like not the case for you.

Reply:
I have no issues running the build on clean box (Docker):

```
$ docker run -it gradle:8.7.0-jdk11 sh                                                                                                                                                                                                                                                                        
# git clone https://github.com/opensearch-project/opensearch-java/                                                                                                                                                                                                                                                                                    
Cloning into 'opensearch-java'...                                                                                                                                                                                                                                                                                                                     
remote: Enumerating objects: 64193, done.        
remote: Counting objects: 100% (1086/1086), done.                                                                                                                                                                                                                                                                                                     
remote: Compressing objects: 100% (530/530), done.
remote: Total 64193 (delta 410), reused 926 (delta 292), pack-reused 63107                                                                                                                                                                                                                                                                            
Receiving objects: 100% (64193/64193), 26.29 MiB | 7.66 MiB/s, done.
Resolving deltas: 100% (54268/54268), done.
# cd opensearch-java        
# ./gradlew check           
Downloading https://services.gradle.org/distributions/gradle-8.7-bin.zip
............10%.............20%.............30%.............40%............50%.............60%.............70%.............80%.............90%............100%             
Starting a Gradle Daemon (subsequent builds will be faster)
                                                                                     
> Configure project :java-client                                                                                                                                           
=======================================                                                                                                                                                                                                                                                                                                               
  Runtime JDK Version   : 11                                                                                                                                               
  Gradle JDK Version    : 11                                                                                                                                                                                                                                                                                                                          
=======================================                                                                                                                                    
                                                                                                                                                                                                                                                                                                                                                      
> Task :java-client:compileJava                                                                                                                                                                                                                                                                                                                       
Note: Some input files use or override a deprecated API.                                                                                                                   
Note: Recompile with -Xlint:deprecation for details.                                                                                                                       
Note: Some input files use unchecked or unsafe operations.                                                                                                                 
Note: Recompile with -Xlint:unchecked for details.                                                                                                                         
                                                                                     
> Task :java-client:compileTestJava                                                                                                                                        
Note: Some input files use or override a deprecated API.                                                                                                                                                                                                                                                                                              
Note: Recompile with -Xlint:deprecation for details.                                                                                                                                                                                                                                                                                                  
                                                                                     
> Task :samples:compileJava                                                                                                                                                                                                                                                                                                                           
Note: /home/gradle/opensearch-java/samples/src/main/java/org/opensearch/client/samples/Search.java uses unchecked or unsafe operations.                                                                                                                                                                                                               
Note: Recompile with -Xlint:unchecked for details.                                                                                                                                                                                                                                                                                                    
                                                                                     
Deprecated Gradle features were used in this build, making it incompatible with Gradle 9.0.                                                                                
                                                                                                                                                                           
You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.                                                                                                                                                                                                             

For more on this, please refer to https://docs.gradle.org/8.7/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.
                                                                                                                                                                                                                                                                                                                                                      
BUILD SUCCESSFUL in 2m 24s                                                            
14 actionable tasks: 14 executed                                                      

```

Reply:
I didn't change anything, but now it builds fine again. 
I think there may have been a temporary issue with https://aws.oss.sonatype.org/content/repositories/snapshots.
I'll have to look into this a bit more, but in any case, I'm back to work.